File: learning-one-neuron.go
package main
import "fmt"


var problem = [] string {
    "...?.....?............#",
    "......?.......?....####",
    ".?........?.....#####?#",
    ".......?.....#####?####",
    "..?..?....#############",
    ".......######?######?##",
    "....#####?#######?#####",
    ".###################?##" }


var wx = 0.0
var wy = 0.0
var wb = 0.0

func neuron(x, y int) int {
    var bias=1.0
    var f = wx * float64(x) + wy * float64(y) + wb * bias
    if f < 0 {
        return -1
    }
    return 1
}

func learn(x, y, desired int) {
	var learning_constant = 0.1
	var error = desired - neuron(x, y)
	wx += float64(error * x) * learning_constant
	wy += float64(error * y) * learning_constant
	wb += float64(error * 1) * learning_constant  // bias = 1
}

func one_step_of_learning() {
	for y, line := range problem {
		for x, ch := range line {
			if (ch == '.') {
				learn(x, y, -1)
			}
			if (ch == '#') {
				learn(x, y, 1)
			}
		}
	}
}

func print_solution() {
	for y, line := range problem {
		var new_line string
		for x, ch := range line {
			var char_ = ch  // char_ := ch
			if ch == '?' {
				var r = neuron(x, y)
				if r < 0 {
					char_ = '!'
				} else {
					char_ = '%'
				}
			}
			// maybe it will be better with use  var bb bytes.Buffer , or  var sb strings.Builder
			new_line += string(char_)
		}
		fmt.Println(new_line)
	}
}

func main() {
	print_solution()
	for i := 1; i < 100; i++ {
        one_step_of_learning()
		print_solution()
    }
}

File: learning_one_neuron.pas
program learning_one_neuron;
const
   width_lim : integer = 23;
   height_lim : integer = 7;

   problem : array [0..7, 0..23] of char = (
      '...?.....?............#',
      '......?.......?....####',
      '.?........?.....#####?#',
      '.......?.....#####?####',
      '..?..?....#############',
      '.......######?######?##',
      '....#####?#######?#####',
      '.###################?##');

var
   wx : real = 0;
   wy : real = 0;
   wb : real = 0;
   (* bias : integer = 1; *)
   i : integer;

   function neuron(x, y : integer): integer; overload;  (* (x, y : integer; bias: integer = 1) *)
   var
      f : real;
      bias : integer;
   begin
      bias := 1;
      f := wx * x + wy * y + wb * bias;
      if (f < 0) then
         neuron := -1
      else
         neuron := 1;
   end;

   function neuron(x, y, bias : integer): integer; overload;
   var
      f : real;
   begin
      f := wx * x + wy * y + wb * bias;
      if (f < 0) then
         neuron := -1
      else
         neuron := 1;
   end;

   procedure learn(x, y, desired : integer);
   const
      learning_constant : real = 0.1;
   var
      error : integer;
   begin
      error := desired - neuron(x, y);
      wx := wx + error * x * learning_constant;
      wy := wy + error * y * learning_constant;
      wb := wb + error * 1 * learning_constant;  (* bias = 1 *)
   end;

   procedure one_step_of_learning;
   var
      y : integer;
      x : integer;
      char_ : char;
   begin
      for y := 0 to height_lim do
      begin
         for x := 0 to width_lim do
         begin
            char_ := problem[y][x];
            if (char_ = '.') then
               learn(x, y, -1);
            if (char_ = '#') then
               learn(x, y, 1);
         end;
      end;
   end;

   procedure print_solution;
   var
      y : integer;
      x : integer;
      char_ : char;
      new_line : array [0..23] of char;
      r : integer;
   begin
      for y := 0 to height_lim do
      begin
         for x := 0 to width_lim do
         begin
            char_ := problem[y][x];
            if (char_ = '?') then
            begin
               r := neuron(x, y);
               if (r < 0) then
                  char_ := '!'
               else
                  char_ := '%';
            end;
            new_line[x] := char_;
         end;
         writeln(new_line);
      end;
   end;

begin
   print_solution;
   for i := 0 to 99 do
   begin
      one_step_of_learning;
      print_solution;
   end;
end.

